### PR TITLE
Run the agent as a Windows service.

### PR DESCRIPTION
**Pull request by vkalintiris**
##### Summary

This commit contains the boilerplate code for running the agent as a Windows service.

To create an installer you should run:

1. `./packaging/utils/compile-on-windows.sh`, and
2. `./packaging/utils/package-windows.sh`

You can then install the agent, which should create a Netdata service and start it automatically.

You can run the agent as a regular executable as well, by running it from the terminal.

##### Test Plan

- CI jobs.

### CODE REVIEW

**Comment by thiagoftsm** [comment]
> However, at this point we need to resolve the issue of gaps when running the agent as a service. It seems that sleeping for one second with sleep(1), actually sleeps for 2 to 4 seconds on my setup.

This is expected. Now that we are working with Microsoft, we need to use more deep their own API, for example:

- We have [Sleep](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) function that probably calls [SleepEx](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleepex).
- We can also use [CreateWaitable](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createwaitabletimerw?redirectedfrom=MSDN)

When we call some normal function that we use on Linux, in normal situation it is converted to some internal windows function. 

We also need to have in our minds that Microsoft does not measure time like Unix. They update internal counters each [100 ns](https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times).

**Comment by thiagoftsm** [comment]
> Once we resolve this, the work that remains concerns packaging: ie. installing the binaries at the proper places so that the relevant DLLs are found.

Please, do not forget to require administrative privileges to install the Service. I remember some installers have option to do this.

**Comment by thiagoftsm** [comment]
> (Only once) Create the netdata service: sc.exe config Netdata binPath="C:\msys64\usr\bin\netdata"

I still did not study the source code, but I suggest changing this. Let us add two new options:

- --install-service: This option will be used by installer for netdata register itself as a Service.
- --remove-service: Remove netdata service.

If it is planned to remove and install the service with an installer, this suggestion can be ignored.

**Comment by thiagoftsm** [comment]
> Last time I checked, the agent crashes after a while when using dbengine. In order to have something that works correctly, you should specify memory-mode ram in your netdata.conf.

This probably has relationship with threads or permissions. I will run netdata with debugger to check.

File: src/daemon/main.c [hunk 1]
```diff
@@ -1411,7 +1411,12 @@ int unittest_prepare_rrd(char **user) {
     return 0;
 }
 
-int main(int argc, char **argv) {
+#ifdef OS_WINDOWS
+int netdata_main(int argc, char **argv)
```
Comment (thiagoftsm) [inline src/daemon/main.c:1415]: A comment: I understand we are still not using a [WinMain](https://learn.microsoft.com/en-us/windows/win32/learnwin32/winmain--the-application-entry-point) as an entry point, but we will have it one day.

File: src/daemon/winsvc.cc [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+static ND_THREAD *main_thread = nullptr;
+static SERVICE_STATUS_HANDLE svc_status_handle = nullptr;
+
+#ifdef NETDATA_INTERNAL_CHECKS
+static void WriteLog(const std::string &message)
+{
+    SYSTEMTIME time;
+    GetSystemTime(&time);
+
+    std::ofstream log("/opt/netdata/var/log/netdata/service.log", std::ios_base::app);
```
Comment (thiagoftsm) [inline src/daemon/winsvc.cc:18]: Probably we should start using `C:\\Program Files\\Netdata`. :thinking:

**Comment by thiagoftsm** [comment]
I thought this could be the scenario, so I decided to test:

Screenshot_20240529_133008

When we run direct on Windows, ignoring `msys`, we have an error.

**Comment by thiagoftsm** [comment]
@vkalintiris , please, change :

```sh
sc.exe config Netdata binPath="C:\msys64\usr\bin\netdata"
```

to

```sh
sc.exe create "Netdata" binpath="C:\msys64\usr\bin\netdata"
```

**Comment by thiagoftsm** [comment]
When I run this `sc.exe create "Netdata" binpath="C:\msys64\usr\bin\netdata"`, the service does not initiate. I receive error `1053`. According to `Event Viewer`, the service did not start in 30 seconds, independent of memory mode.

**Comment by vkalintiris** [comment]
> > However, at this point we need to resolve the issue of gaps when running the agent as a service. It seems that sleeping for one second with sleep(1), actually sleeps for 2 to 4 seconds on my setup.
> 
> This is expected. Now that we are working with Microsoft, we need to use more deep their own API, for example:
> 
> * We have [Sleep](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) function that probably calls [SleepEx](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleepex).
> * We can also use [CreateWaitable](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createwaitabletimerw?redirectedfrom=MSDN)
> 
> When we call some normal function that we use on Linux, in normal situation it is converted to some internal windows function.
> 
> We also need to have in our minds that Microsoft does not measure time like Unix. They update internal counters each [100 ns](https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times).

I'll check the timings of these functions, but I don't see how `sleep()` could digress so much because it should use one of those behind the scenes.

**Comment by vkalintiris** [comment]
> > (Only once) Create the netdata service: sc.exe config Netdata binPath="C:\msys64\usr\bin\netdata"
> 
> I still did not study the source code, but I suggest changing this. Let us add two new options:
> 
> * --install-service: This option will be used by installer for netdata register itself as a Service.
> * --remove-service: Remove netdata service.
> 
> If it is planned to remove and install the service with an installer, this suggestion can be ignored.

The installer will be an executable that someone can download and start. The installer will be responsible for installing the uninstaller as well in order to allow people to uninstall the agent if they wish.

File: src/daemon/winsvc.cc [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+#if 1
+int main() {
+    SERVICE_TABLE_ENTRY serviceTable[] = {
+        { strdupz("Netdata"), ServiceMain },
+        { nullptr, nullptr }
+    };
+
+    if (!StartServiceCtrlDispatcher(serviceTable))
+    {
+        WriteLog("@main() - StartServiceCtrlDispatcher() failed...");
```
[the remark this answers is not in the document]
Comment (vkalintiris) [reply src/daemon/winsvc.cc:132]: We want to create the service with `sc.exe` at installation time, instead of adding this logic in the agent itself.

**Comment by vkalintiris** [comment]
> I thought this could be the scenario, so I decided to test:
> 
> Screenshot_20240529_133008
> 
> When we run direct on Windows, ignoring `msys`, we have an error.

A feature not a bug, at this point.

**Comment by vkalintiris** [comment]
> When I run this `sc.exe create "Netdata" binpath="C:\msys64\usr\bin\netdata"`, the service does not initiate. I receive error `1053`. According to `Event Viewer`, the service did not start in 30 seconds, independent of memory mode.

`sc.exe create ...` creates the service, it should not start it. Did you mean to say something else?

**Comment by vkalintiris** [comment]
> > > However, at this point we need to resolve the issue of gaps when running the agent as a service. It seems that sleeping for one second with sleep(1), actually sleeps for 2 to 4 seconds on my setup.
> > 
> > 
> > This is expected. Now that we are working with Microsoft, we need to use more deep their own API, for example:
> > 
> > * We have [Sleep](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleep) function that probably calls [SleepEx](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-sleepex).
> > * We can also use [CreateWaitable](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createwaitabletimerw?redirectedfrom=MSDN)
> > 
> > When we call some normal function that we use on Linux, in normal situation it is converted to some internal windows function.
> > We also need to have in our minds that Microsoft does not measure time like Unix. They update internal counters each [100 ns](https://learn.microsoft.com/en-us/windows/win32/sysinfo/file-times).
> 
> I'll check the timings of these functions, but I don't see how `sleep()` could digress so much because it should use one of those behind the scenes.

Apparently, on Windows the highest resolution sleep is 1 msec. We use various kind of sleep functions throughout the entire codebase. For time being, I've only updated the sleep function that is called within the `heartbeat_next` function. I don't notice any gaps in the charts now. However, other places in the agent might need to be updated as well, in the future.

**Comment by StefanSa** [comment]
@vkalintiris, @ktsaou  Hi there,
I had the service startup problem with exactly the same error message, currently i solved this with a service wrapper.
I have created an installer for netdata with InnoSetup.
It installs netdata and the necessary dll's regardless of the file path, so you don't have to install msys2.
The netdata service is also installed. The uninstall routine also works without any problems.
Currently i am still optimizing for updates with InnoSetup and netdata.
If you like, i can provide you with the installation script of InnoSetup.

**Review by thiagoftsm — CHANGES_REQUESTED** [review]
Considering I am not unique to have [issues](https://github.com/netdata/netdata/pull/17766#issuecomment-2143780288) and considering that we also have issues to do cross-compilation due `pthread`, I am doing this `request change` to avoid merge this PR until we address all issues.

**Comment by thiagoftsm** [comment]
> When I run this `sc.exe create "Netdata" binpath="C:\msys64\usr\bin\netdata"`, the service does not initiate. I receive error `1053`. According to `Event Viewer`, the service did not start in 30 seconds, independent of memory mode.

Hello @vkalintiris ,

I still have the same issue. I am doing the following steps to have it:

##  On Windows prompt
- I start a terminal as administrator and I ran: `sc delete "Netdata"`

## On Msys2
- I clone your branch
- I compiled it

## On Windows prompt
-  I ran: sc.exe create "Netdata" binpath="C:\msys64\opt\netdata\usr\sbin\netdata.exe"

## On Windows
- I went to Start -> `run` and I wrote `services.msc`
- I found the `Netdata` service installed, and I tried to start it.

Best regards.

Thiago

File: src/libnetdata/locks/locks.h [hunk 4]
```diff
@@ -9,16 +9,27 @@
...
 typedef struct netdata_spinlock {
     bool locked;
 #ifdef NETDATA_INTERNAL_CHECKS
     pid_t locker_pid;
     size_t spins;
 #endif
 } SPINLOCK;
+#else
+typedef struct netdata_spinlock {
+    netdata_mutex_t inner;
```
Comment (thiagoftsm) [inline src/libnetdata/locks/locks.h:22]: When we cross compile our code on Linux, `netdata_mutex_t` is one of the issues raised. Is there the possibility we already start using Windows native functions and data type? Or to use `libuv`?

**Comment by vkalintiris** [comment]
@thiagoftsm @stelfrag Please, check the PR's description if you want to give it a try. Post any errors you might find here.

**Comment by vkalintiris** [comment]
@thiagoftsm @stelfrag The latest commit disables pluginsd. Without it the agent starts up correctly all the time. AFAIC, this unblocks the Windows release, considering that we don't rely on pluginsd for collecting Windows metrics. Of course, I'll keep investigating until I find the real, underlying cause of the problem.

**Comment by StefanSa** [comment]
@vkalintiris,  Thanks for your time and effort.
but i can't get the agent from your WIndows branch to work
`git clone -b windows https://github.com/vkalintiris/netdata.git`
`git submodule init && git submodule update`
`packaging/utils/compile-on-windows.sh`

Copy the necessary dlls to it, but when starting in the cmd box nothing is displayed at all,
no error message can be seen in the event list either.
It starts for a brief second and then stops.
This works in the original repo without any problem, except that the agent does not run as a service and you have to use a service wrapper. Any idea where the problem could be?

**Comment by vkalintiris** [comment]
> @vkalintiris,  Thanks for your time and effort.
> but i can't get the agent from your WIndows branch to work
> `git clone -b windows https://github.com/vkalintiris/netdata.git`
> `git submodule init && git submodule update`
> `packaging/utils/compile-on-windows.sh`
> 
> Copy the necessary dlls to it, but when starting in the cmd box nothing is displayed at all,
> no error message can be seen in the event list either.
> It starts for a brief second and then stops.
> This works in the original repo without any problem, except that the agent does not run as a service and you have to use a service wrapper. Any idea where the problem could be?

I have updated the PR description. You need to create the installer and setup the service for this to work.

**Comment by thiagoftsm** [comment]
Hello @StefanSa ,

When you talk about `cmd`, I understand this is the windows terminal instead `msys` terminal, please, can you confirm this?

I have tested this PR on `msys` and also on windows terminal, and I observed fails when I am running on the last.

I will test now latest commits.

Best Regards,

Thiago!

File: src/daemon/winsvc.cc [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    log << time.wHour << ":" << time.wMinute << ":" << time.wSecond << " - " << message << std::endl;
+}
+#else
+static void WriteLog(const std::string &message)
+{
+    UNUSED(message);
+}
+#endif
+
+static void *call_netdata_main(void *arg)
```
Comment (thiagoftsm) [inline src/daemon/winsvc.cc:36]: This static function is not used.

**Comment by thiagoftsm** [comment]
@vkalintiris ,

After last commits, when I try to start the daemon running the command on terminal, I observe that `netdata` was installed in `/opt/netdata`, but the daemon is trying to write logs in `/var/log/netdata`. Of course, this is not expected, so it `daemon` ends with error 1.

**Comment by thiagoftsm** [comment]
@vkalintiris ,

Please, add a description to netdata service:

missingdesc

I could install using our installer, but it fails when I used manual commands as I described previously. If the goal is always to use the installer, we will have to add messages  (I am not saying logs, I am talking about [MessageBoxes](https://learn.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-messageboxexa)) to give proper information for our users;

**Comment by thiagoftsm** [comment]
When I try to stop the service, I have error `1067`:  "Netdata service end was unexpected"

**Comment by thiagoftsm** [comment]
Our installer is not checking a previous netdata installed, this should not happen. If we have a netdata service installed, it should report there is a previous version, or it would need to say that we are going to update. We are repeating the same steps and we try to reinstall the service..

**Comment by thiagoftsm** [comment]
We have the `Unistaller` inside `C:\Program Files\netdata`, but we do not have any information inside `Control Panel - > Software`. Users should also be able to remove netdata from there.

**Comment by StefanSa** [comment]
@thiagoftsm 

Yes, i have only tested this in the cmd terminal,
because i do not want to install msys2 with the
it is sufficient to copy only the necessary msys2 dll. I also see the error 1067 when i try to install the service with my installer.
As already described, i have no problems with the original metadata branch, except that i currently have to use a service wrapper because the native agent does not start as a Service.

**Comment by thiagoftsm** [comment]
@StefanSa, thank you to confirm, and test this PR.

@vkalintiris now we have a confirmation that is not something related to a specific environment.

**Comment by thiagoftsm** [comment]
When we try to use `sc.exe` on windows command line we still have issues to start service. The error is `1053`, but there is a difference this time, my Service controller this time was frozen.

When I ran the new installer created with OP instructions:, I had the following scenario:

- It did not give me option to select destination and it removed completely my development environment on Microsoft.
- It did not create the service.

Finally, I do not know why we changed the installer screen, but I think the previous were more user-friendly.

**Comment by vkalintiris** [comment]
@thiagoftsm give the PR another round of tests (test only the installer and/or sc.exe start/stop/delete, ie. not RUN_FROMM_CLI). AFAICT, it should be working now. However, there are still some issue we'll have to address.

**Comment by thiagoftsm** [comment]
@vkalintiris ,

After last commits, I am not able to make the installer. I am having this error:

Commits

**Comment by vkalintiris** [comment]
> @vkalintiris ,
> 
> After last commits, I am not able to make the installer. I am having this error:
> 
> Commits

You installed msys2 under `C:\msys64_dev`, the packaging script expects the default installation path, ie. `C:\msys64`.

**Comment by StefanSa** [comment]
@vkalintiris 
i can confirm that your last commit works with the installer and service.
But what i notice is the following. A start in the msys2 or cmd terminal fails without any hint / message in the terminal directly after the start. It seems that this version can only be started as a service. How can i tell the service / agent the path of the netdtata.conf (-c), or where does the current agent get its information from which netdata.conf (usr\lib\netdata\conf.d) ?
One more important thing, the installer does not make an entry in "Programs and Features", so that it cannot be uninstalled via this. Likewise, not everything is uninstalled, empty folders remain, e.g. (netdata\usr\sbin).

**Comment by thiagoftsm** [comment]
> You installed msys2 under `C:\msys64_dev`, the packaging script expects the default installation path, ie. `C:\msys64`.

I think we need to change our script to get current local directory. I had to create the environment without default targets, because last time I used the installer, it overwrites everything.

**Comment by ktsaou** [comment]
Guys, I need this to continue working on windows.
Please review and merge asap.

**Comment by StefanSa** [comment]
Hi there,
can now confirm that the netdata agent also works in the terminal.

i still have a few thoughts about the installer:
- The installer should be able to perform updates when the version number increases.
  To do this, the service must be stopped beforehand, otherwise the update will fail.

- Package name should contain version name and platform, e.g. netdata-1.46.0.446-x64.exe
  
  image

- It would be important to specify the path of the config file in the service using "-c".
  However, this would require a placeholder / variable in the config file, which the installer fills in with the correct path.

- The installer should make an entry in "Programs and Features", so that an uninstallation can be executed here.

  image

- When removing the app, no files or folders should be left behind, everything must be uninstalled.

- Asking the user whether the service should be started after installation or stopped during the update increases the user 
  experience.

- It would be nice if the installer showed the Netdata logo.

  image

**Comment by thiagoftsm** [comment]
@vkalintiris ,

Everything said by @StefanSa is right, I do not have nothing to add.

> When removing the app, no files or folders should be left behind, everything must be uninstalled.
On my environment I observed we still have `/usr` and `/var` kept. 

Considering everything critical was fixed. I am going to approve it for we unblock @ktsaou and also myself, but we cannot forgot what @StefanSa mentioned. 

Best regards!

**Review by thiagoftsm — APPROVED** [review]
I am approving because there is not more critical issues, but, again, the minor issues must be address with another PR.

**Comment by vkalintiris** [comment]
@StefanSa I'll address most of your suggestions in follow up PRs. We had to merge this to unblock the rest of the team.

**Pull request merged**